File: frontend/equipment.js
"use strict";
const Fleet = (() => {
  let access = {permissions: []};
  let selected = null;
  let tabName = 'overview';
  let photoUrls = [];
  function clearPhotos(){for(const url of photoUrls)URL.revokeObjectURL(url);photoUrls=[];}
  $('detail-dialog').addEventListener('close',clearPhotos);
  const can = code => access.superadmin || access.permissions.includes(code) || ({'assets.meter.record':'assets.record_meter','assets.documents.read':'asset_documents.read','assets.documents.manage':'asset_documents.manage'}[code] && access.permissions.includes({'assets.meter.record':'assets.record_meter','assets.documents.read':'asset_documents.read','assets.documents.manage':'asset_documents.manage'}[code]));
  const label = key => key.replaceAll('_',' ').replaceAll('-',' ');
  function button(title, task, primary=false) { const node=el('button',title,primary?'primary':'secondary');node.type='button';node.onclick=()=>action(task);return node; }
  const config = {
    assignments:['Assignments','assets.read','assets.assign','AssetAssignmentCreate','asset-assignments'],
    'location-history':['Location history','assets.location.read','assets.location.manage','LocationEvent','asset-location-events'],
    'meter-readings':['Meter readings','assets.meter.read','assets.meter.record','AssetMeterReadingCreate','asset-meter-readings'],
    components:['Components','assets.components.read','assets.components.manage','AssetComponentCreate','asset-components','ComponentUpdate'],
    inspections:['Inspections','assets.inspections.read','assets.inspections.manage','InspectionCreate','asset-inspections','InspectionUpdate'],
    defects:['Defects','assets.defects.read','assets.defects.manage','DefectCreate','asset-defects','DefectUpdate'],
    documents:['Documents','assets.documents.read','assets.documents.manage','AssetDocumentCreate','asset-documents','DocumentUpdate'],
    insurance:['Insurance','assets.insurance.read','assets.insurance.manage','InsuranceCreate','asset-insurance','InsuranceUpdate'],
    registrations:['Registration','assets.registration.read','assets.registration.manage','RegistrationCreate','asset-registrations','RegistrationUpdate'],
    media:['Photos','assets.media.read','assets.media.manage','MediaCreate','asset-media','MediaUpdate'],
    ownership:['Ownership','assets.ownership.read','assets.ownership.manage','OwnershipCreate','asset-ownership'],
    'status-history':['Status history','assets.read'],
    activity:['Activity','assets.audit.read']
  };
  async function initialize(){access=await api('/api/v1/equipment/access');}
  function form(title,schema,path,opts={}){
    const lookups={category_id:'asset-categories',parent_category_id:'asset-categories',default_location_id:'locations',responsible_employee_id:'employees',primary_operator_id:'employees',inspected_by_id:'users',parent_component_id:`assets/${selected}/components`,document_id:`assets/${selected}/documents`,inspection_id:`assets/${selected}/inspections`};
    return Workforce.form(title,schema,path,{contextId:null,lookupTargets:lookups,done:()=>selected?open(selected,tabName):loadList('assets'),...opts});
  }
  async function create(){await initialize();selected=null;return form('Register equipment','AssetCreate','/api/v1/assets',{done:async row=>{await loadList('assets');await open(row.id);}});}
  function facts(data,parent){const list=el('dl',undefined,'detail-list');for(const [key,value] of Object.entries(data)){if(value==null||['id','organization_id','asset_id','storage_path','created_by_id','updated_by_id'].includes(key)||typeof value==='object')continue;list.append(el('dt',label(key)),el('dd',String(value)));}parent.append(list);}
  async function open(id, tab='overview'){
    clearPhotos();selected=id;tabName=tab;await initialize();
    const overview=await api(`/api/v1/assets/${id}/overview`);
    openDialogTitle('EQUIPMENT · '+overview.asset.asset_number,overview.asset.name);
    $('detail-dialog').classList.add('equipment-dialog');
    const area=$('detail-body');
    const banner=el('div',undefined,'equipment-banner');
    banner.append(el('strong',overview.status),el('span',overview.operational_eligibility),el('span',overview.current_project?.name||'No active project'),el('span',overview.current_location?.name||'Location not recorded'));
    area.append(banner);
    const tabs=el('nav',undefined,'equipment-tabs');tabs.setAttribute('aria-label','Equipment record sections');
    for(const [key,title] of [['overview','Overview'],...Object.entries(config).filter(([,cfg])=>can(cfg[1])).map(([key,cfg])=>[key,cfg[0]])]){const b=button(title,()=>open(id,key));b.setAttribute('aria-pressed',String(key===tab));tabs.append(b);}area.append(tabs);
    const content=el('div',undefined,'equipment-content');area.append(content);
    if(tab==='overview'){
      const actions=el('div',undefined,'row-actions');
      if(can('assets.update'))actions.append(button('Edit equipment',()=>form('Edit equipment','AssetUpdate',`/api/v1/assets/${id}`,{record:overview.asset})));
      if(can('assets.status.change'))actions.append(button('Change status',()=>form('Change operational status','StatusChange',`/api/v1/assets/${id}/status`)));
      if(can('assets.transfer')&&overview.current_assignment)actions.append(button('Transfer equipment',()=>form('Transfer equipment','AssetTransfer',`/api/v1/assets/${id}/transfer`)));
      if(can('assets.archive'))actions.append(button(overview.asset.is_active?'Archive':'Restore',async()=>{await api(`/api/v1/assets/${id}/${overview.asset.is_active?'archive':'restore'}`,{method:'POST'});await open(id);}));
      content.append(actions);
      for(const reason of overview.reasons)content.append(el('p',reason,'access-note'));
      facts({...overview.asset,category:overview.category?.name,project:overview.current_project?.name,location:overview.current_location?.name,responsible:overview.responsible_employee&&`${overview.responsible_employee.first_name} ${overview.responsible_employee.last_name}`,operator:overview.primary_operator&&`${overview.primary_operator.first_name} ${overview.primary_operator.last_name}`,meter_status:overview.meter_status,meter_reading_age_days:overview.meter_reading_age_days},content);
      return;
    }
    const cfg=config[tab];
    const toolbar=el('div',undefined,'row-actions');content.append(toolbar);
    if(cfg[2]&&can(cfg[2]))toolbar.append(button('Add '+cfg[0].toLowerCase(),()=>form('Add '+cfg[0].toLowerCase(),cfg[3],`/api/v1/assets/${id}/${tab==='location-history'?'location-events':tab}`),true));
    if(tab==='meter-readings'&&can('assets.meter.override'))toolbar.append(button('Reset / replace meter',()=>form('Reset or replace meter','MeterReset',`/api/v1/assets/${id}/meter-reset`)));
    if(tab==='documents'&&can(cfg[2]))content.append(uploadForm('assets',id));
    if(tab==='media'&&can(cfg[2])){
      const input=el('input');input.type='file';input.accept='.png,.jpg,.jpeg';input.setAttribute('aria-label','Choose equipment photo');
      const error=el('p','','form-error');error.id='equipment-photo-error';
      toolbar.append(input,button('Upload photo',async()=>{if(!input.files[0])throw Error('Choose a photo');const body=new FormData();body.append('file',input.files[0]);await sendForm(`/api/v1/assets/${id}/media/upload`,body);await open(id,tab);}));content.append(error);
    }
    try{
      const response=await api(`/api/v1/assets/${id}/${tab}`);const rows=Array.isArray(response)?response:response.items;
      if(!rows.length)content.append(el('p','No records yet.','hint'));
      for(const row of rows){
        const card=el('article',undefined,'employee-record');facts(row,card);const actions=el('div',undefined,'row-actions');card.append(actions);content.append(card);
        if(cfg[5]&&can(cfg[2]))actions.append(button('Edit',()=>form('Edit '+cfg[0],cfg[5],`/api/v1/${cfg[4]}/${row.id}`,{record:row})));
        if(tab==='assignments'&&['ACTIVE','PLANNED'].includes(row.status)&&can('assets.assign')){
          actions.append(button('Complete',()=>form('Complete assignment','AssetAssignmentUpdate',`/api/v1/asset-assignments/${row.id}/complete`)),button('Cancel assignment',async()=>{await api(`/api/v1/asset-assignments/${row.id}/cancel`,{method:'POST'});await open(id,tab);}));
        }
        if(tab==='components'&&can(cfg[2])&&!['REMOVED','REPLACED','SCRAPPED'].includes(row.status))actions.append(button('Remove',()=>form('Remove component','ComponentRemoval',`/api/v1/asset-components/${row.id}/remove`)),button('Replace',()=>form('Install replacement component','AssetComponentCreate',`/api/v1/asset-components/${row.id}/replace`)));
        if(tab==='defects'&&can(cfg[2])&&['OPEN','ACKNOWLEDGED'].includes(row.status))actions.append(button('Resolve',()=>form('Resolve defect','ResolveDefect',`/api/v1/asset-defects/${row.id}/resolve`)));
        if(tab==='documents'){
          if(row.file_url?.startsWith('/api/'))actions.append(downloadButton('Download document',row.file_url));
          else if(/^https?:\/\//.test(row.file_url||'')){const link=el('a','Open document','secondary');link.href=row.file_url;link.target='_blank';link.rel='noopener noreferrer';actions.append(link);}
          for(const op of ['verify','archive'])if(can(op==='verify'?'assets.documents.verify':cfg[2])&&row.is_active)actions.append(button(label(op),async()=>{await api(`/api/v1/asset-documents/${row.id}/${op}`,{method:'POST'});await open(id,tab);}));
        }
        if(tab==='media'){
          if(row.media_type==='PHOTO'&&row.is_active&&row.file_url?.startsWith('/api/')){const result=await apiBlob(row.file_url);const image=el('img');image.alt=row.caption||row.file_name||'Equipment photo';image.className='equipment-photo';image.src=URL.createObjectURL(result.blob);photoUrls.push(image.src);card.prepend(image);}
          if(row.file_url?.startsWith('/api/'))actions.append(downloadButton('Download photo',row.file_url));
          if(can(cfg[2])&&row.is_active)for(const op of ['set-primary','archive'])actions.append(button(label(op),async()=>{await api(`/api/v1/asset-media/${row.id}/${op}`,{method:'POST'});await open(id,tab);}));
        }
      }
    }catch(error){content.append(el('p',error.message,'form-error'));}
  }
  async function dashboard(){
    await initialize();
    let area=$('fleet-summary');
    if(!area){area=el('div');area.id='fleet-summary';$('assets-view').insertBefore(area,$('assets-view').children[1]);}
    area.replaceChildren();
    const data=await api('/api/v1/assets/dashboard-summary');const summary=el('div',undefined,'equipment-summary');
    for(const key of ['total_assets','available_assets','assigned_assets','maintenance_assets','breakdown_assets','critical_open_defects']){const item=el('article');item.append(el('strong',String(data[key]??0)),el('span',label(key)));summary.append(item);}area.append(summary);
    const tools=el('div',undefined,'row-actions');
    if(can('assets.create'))tools.append(button('New category',()=>form('New asset category','AssetCategoryCreate','/api/v1/asset-categories',{done:()=>loadList('assets')})));
    tools.append(button('Manage categories',categories));
    const code=el('input');code.placeholder='Scan or enter asset tag';code.setAttribute('aria-label','Asset tag lookup');tools.append(code,button('Find asset tag',async()=>{const result=await api('/api/v1/assets/lookup?code='+encodeURIComponent(code.value.trim()));await open(result.asset.id);}));area.append(tools);
    if(!$('fleet-category_id')){const cats=await api('/api/v1/asset-categories');for(const [key,title,choices] of [['category_id','All categories',cats.map(c=>[c.id,c.name])],['status','All statuses',['AVAILABLE','ASSIGNED','OPERATING','STANDBY','MOBILIZING','DEMOBILIZING','UNDER_MAINTENANCE','BREAKDOWN','OUT_OF_SERVICE','QUARANTINED','DISPOSED','LOST','STOLEN'].map(v=>[v,label(v)])]]){const select=el('select');select.id='fleet-'+key;select.setAttribute('aria-label',title);select.append(new Option(title,''));for(const [value,name] of choices)select.append(new Option(name,value));select.onchange=()=>action(()=>loadList('assets'));$('asset-search').parentElement.prepend(select);}}
    if(!$('fleet-filter')){const select=el('select');select.id='fleet-filter';select.setAttribute('aria-label','Filter equipment');for(const [value,title] of [['','All equipment'],['is_available=true','Available for deployment'],['has_critical_defects=true','Critical defects'],['document_expiring_within_days=30','Documents expiring within 30 days'],['insurance_expiring_within_days=30','Insurance expiring within 30 days'],['registration_expiring_within_days=30','Registration expiring within 30 days']])select.append(new Option(title,value));select.onchange=()=>action(()=>loadList('assets'));$('asset-search').parentElement.prepend(select);}
  }
    async function categories(){openDialogTitle('EQUIPMENT','Asset categories');const rows=await api('/api/v1/asset-categories');for(const row of rows){const card=el('article',undefined,'employee-record');facts(row,card);if(can('assets.update'))card.append(button('Edit category',()=>form('Edit category','CategoryUpdate',`/api/v1/asset-categories/${row.id}`,{record:row,done:categories})));$('detail-body').append(card);}}
  function filters(){let result=$('fleet-filter')?.value?'&'+$('fleet-filter').value:'';for(const key of ['category_id','status'])if($('fleet-'+key)?.value)result+='&'+key+'='+encodeURIComponent($('fleet-'+key).value);return result;}
  return {open,create,dashboard,filters};
})();

File: app/api/v1/endpoints/equipment.py
"""Equipment phase routes. Static fleet routes precede /assets/{asset_id}."""

import math
import uuid
from datetime import date, datetime, timedelta
from typing import Any

from fastapi import APIRouter, Depends, File, Form, Query, Request, UploadFile
from fastapi.responses import FileResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from starlette.concurrency import run_in_threadpool

from app.core.dependencies import get_current_active_user, require_permission
from app.core.exceptions import NotFoundError, ValidationError
from app.db.session import get_session
from app.models import (
    Asset,
    AssetAssignment,
    AssetCategory,
    AssetComponent,
    AssetDocument,
    AssetMeterReading,
    User,
)
from app.models.asset import AssetStatus, ComponentStatus, MeterType, OwnershipType
from app.models.asset_records import AssetDefect, AssetMedia
from app.models.employee import AssignmentStatus
from app.schemas import equipment as schemas
from app.schemas.asset import AssetAssignmentUpdate, AssetComponentCreate
from app.services.equipment import RECORDS, EquipmentService

router = APIRouter(tags=["equipment"])


@router.get("/assets")
async def asset_list(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    search: str | None = None,
    category_id: uuid.UUID | None = None,
    status: AssetStatus | None = None,
    ownership_type: OwnershipType | None = None,
    project_id: uuid.UUID | None = None,
    location_id: uuid.UUID | None = None,
    responsible_employee_id: uuid.UUID | None = None,
    primary_operator_id: uuid.UUID | None = None,
    manufacturer: str | None = None,
    model: str | None = None,
    registration_number: str | None = None,
    serial_number: str | None = None,
    is_active: bool | None = None,
    is_available: bool | None = None,
    unassigned_only: bool = False,
    operational_eligibility: str | None = None,
    document_expiring_within_days: int | None = Query(None, ge=0, le=3650),
    insurance_expiring_within_days: int | None = Query(None, ge=0, le=3650),
    registration_expiring_within_days: int | None = Query(None, ge=0, le=3650),
    has_open_defects: bool | None = None,
    has_critical_defects: bool | None = None,
    sort: str = Query("asset_number", pattern="^(asset_number|name|status|manufacturer)$"),
    descending: bool = False,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    filters = locals().copy()
    rows = await EquipmentService(session, actor).filtered_assets(filters)
    rows.sort(key=lambda row: str(row.get(sort) or ""), reverse=descending)
    return {
        "items": rows[(page - 1) * page_size : page * page_size],
        "total": len(rows),
        "page": page,
        "page_size": page_size,
        "pages": math.ceil(len(rows) / page_size),
    }


@router.get("/assets/available")
async def available_assets(
    category_id: uuid.UUID | None = None,
    project_id: uuid.UUID | None = None,
    location_id: uuid.UUID | None = None,
    meter_type: MeterType | None = None,
    ownership_type: OwnershipType | None = None,
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    rows = await EquipmentService(session, actor).filtered_assets(
        {**locals(), "is_available": True}
    )
    return {
        "items": rows[(page - 1) * page_size : page * page_size],
        "total": len(rows),
        "page": page,
        "page_size": page_size,
        "pages": math.ceil(len(rows) / page_size),
    }


@router.get("/assets/lookup")
async def lookup(
    code: str,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    from sqlalchemy import or_

    service = EquipmentService(session, actor)
    row = await session.scalar(
        select(Asset).where(
            Asset.organization_id == actor.organization_id,
            or_(
                Asset.asset_number == code, Asset.qr_code_value == code, Asset.barcode_value == code
            ),
        )
    )
    if row is None:
        raise NotFoundError("Asset not found")
    return await service.overview(row.id)


@router.get("/assets/dashboard-summary")
async def dashboard(
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    return await EquipmentService(session, actor).fleet_summary()


@router.get("/projects/{project_id}/equipment-summary")
async def project_equipment(
    project_id: uuid.UUID,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> dict[str, Any]:
    return await EquipmentService(session, actor).fleet_summary(project_id)


@router.post("/assets/{asset_id}/restore")
async def restore(
    asset_id: uuid.UUID,
    actor: User = Depends(require_permission("assets.archive")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).restore(asset_id)


@router.post("/assets/{asset_id}/status")
async def status(
    asset_id: uuid.UUID,
    body: schemas.StatusChange,
    actor: User = Depends(require_permission("assets.status.change")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).change_status(asset_id, body)


@router.post("/assets/{asset_id}/transfer")
async def transfer(
    asset_id: uuid.UUID,
    body: schemas.AssetTransfer,
    actor: User = Depends(require_permission("assets.transfer")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).transfer(asset_id, body)


@router.post("/assets/{asset_id}/meter-reset")
async def reset(
    asset_id: uuid.UUID,
    body: schemas.MeterReset,
    actor: User = Depends(require_permission("assets.meter.override")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).reset_meter(asset_id, body)


@router.get("/assets/{asset_id}/activity")
async def activity(
    asset_id: uuid.UUID,
    actor: User = Depends(require_permission("assets.audit.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).activity(asset_id)


@router.get("/asset-categories/{identifier}")
async def category(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    return service.public(await service.ref(AssetCategory, identifier))


@router.patch("/asset-categories/{identifier}")
async def category_update(
    identifier: uuid.UUID,
    body: schemas.CategoryUpdate,
    actor: User = Depends(require_permission("assets.update")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).update_category(
        identifier, body.model_dump(exclude_unset=True)
    )


@router.post("/asset-categories/{identifier}/archive")
async def category_archive(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.archive")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).update_category(identifier, {"is_active": False})


@router.get("/asset-assignments/{identifier}")
async def assignment(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    return service.public(await service.ref(AssetAssignment, identifier))


@router.post("/asset-assignments/{identifier}/complete")
async def complete(
    identifier: uuid.UUID,
    body: AssetAssignmentUpdate,
    actor: User = Depends(require_permission("assets.assign")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).update_assignment(
        identifier, body.model_copy(update={"status": AssignmentStatus.COMPLETED})
    )


@router.post("/asset-assignments/{identifier}/cancel")
async def cancel(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.assign")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).update_assignment(
        identifier, AssetAssignmentUpdate(status=AssignmentStatus.CANCELLED)
    )


@router.get("/asset-meter-readings/{identifier}")
async def meter(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.meter.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    return service.public(await service.ref(AssetMeterReading, identifier))


@router.get("/asset-components/{identifier}")
async def component(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.components.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    return service.public(await service.ref(AssetComponent, identifier))


@router.patch("/asset-components/{identifier}")
async def component_update(
    identifier: uuid.UUID,
    body: schemas.ComponentUpdate,
    actor: User = Depends(require_permission("assets.components.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).change_component(
        identifier, body.model_dump(exclude_unset=True)
    )


@router.post("/asset-components/{identifier}/replace")
async def component_replace(
    identifier: uuid.UUID,
    body: AssetComponentCreate,
    actor: User = Depends(require_permission("assets.components.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).change_component(
        identifier, body.model_dump(), True
    )


@router.post("/asset-components/{identifier}/remove")
async def component_remove(
    identifier: uuid.UUID,
    body: schemas.ComponentRemoval,
    actor: User = Depends(require_permission("assets.components.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).change_component(
        identifier, {**body.model_dump(), "status": ComponentStatus.REMOVED, "is_active": False}
    )


def related_routes(
    kind: str, item: str, create_schema: Any, update_schema: Any, permission: str
) -> None:
    async def listing(
        asset_id: uuid.UUID,
        actor: User = Depends(require_permission(permission + ".read")),
        session: AsyncSession = Depends(get_session),
    ) -> Any:
        return await EquipmentService(session, actor).list_records(asset_id, kind)

    listing.__name__ = kind.replace("-", "_") + "_list"
    router.add_api_route("/assets/{asset_id}/" + kind, listing, methods=["GET"])

    async def create(
        asset_id: uuid.UUID,
        body: Any,
        actor: User = Depends(require_permission(permission + ".manage")),
        session: AsyncSession = Depends(get_session),
    ) -> Any:
        return await EquipmentService(session, actor).add_record(asset_id, kind, body)

    create.__annotations__["body"] = create_schema
    create.__name__ = kind.replace("-", "_") + "_create"
    router.add_api_route(
        "/assets/{asset_id}/" + ("location-events" if kind == "location-history" else kind),
        create,
        methods=["POST"],
        status_code=201,
    )
    if update_schema:

        async def patch(
            identifier: uuid.UUID,
            body: Any,
            actor: User = Depends(require_permission(permission + ".manage")),
            session: AsyncSession = Depends(get_session),
        ) -> Any:
            return await EquipmentService(session, actor).update_record(identifier, kind, body)

        patch.__annotations__["body"] = update_schema
        patch.__name__ = kind.replace("-", "_") + "_update"
        router.add_api_route("/" + item + "/{identifier}", patch, methods=["PATCH"])

    async def get(
        identifier: uuid.UUID,
        actor: User = Depends(require_permission(permission + ".read")),
        session: AsyncSession = Depends(get_session),
    ) -> Any:
        service = EquipmentService(session, actor)
        return service.public(await service.ref(RECORDS[kind], identifier))

    get.__name__ = kind.replace("-", "_") + "_get"
    router.add_api_route("/" + item + "/{identifier}", get, methods=["GET"])


def expiry_route(path: str, model: Any, permission: str) -> None:
    async def expiring(
        days: int = Query(30, ge=0, le=3650),
        expired: bool = False,
        actor: User = Depends(require_permission(permission)),
        session: AsyncSession = Depends(get_session),
    ) -> Any:
        service = EquipmentService(session, actor)
        query = select(model).where(
            model.organization_id == actor.organization_id,
            model.expiry_date < date.today()
            if expired
            else model.expiry_date <= date.today() + timedelta(days=days),
        )
        if hasattr(model, "is_active"):
            query = query.where(model.is_active.is_(True))
        elif hasattr(model, "status"):
            query = query.where(model.status.notin_(["CANCELLED", "PENDING"]))
        return [
            service.public(row)
            for row in (await session.scalars(query.order_by(model.expiry_date))).all()
        ]

    expiring.__name__ = path.replace("/", "_").replace("-", "_")
    router.add_api_route(path, expiring, methods=["GET"])


expiry_route("/asset-documents/expiring", AssetDocument, "assets.documents.read")
expiry_route("/asset-insurance/expiring", RECORDS["insurance"], "assets.insurance.read")
expiry_route("/asset-registrations/expiring", RECORDS["registrations"], "assets.registration.read")


@router.get("/asset-defects/open")
@router.get("/asset-defects/critical")
async def defects(
    request: Request,
    actor: User = Depends(require_permission("assets.defects.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    query = select(AssetDefect).where(
        AssetDefect.organization_id == actor.organization_id,
        AssetDefect.status.in_(["OPEN", "ACKNOWLEDGED"]),
    )
    if request.url.path.endswith("/critical"):
        query = query.where(AssetDefect.severity == "CRITICAL")
    service = EquipmentService(session, actor)
    return [
        service.public(row)
        for row in (await session.scalars(query.order_by(AssetDefect.reported_at.desc()))).all()
    ]


related_routes(
    "insurance",
    "asset-insurance",
    schemas.InsuranceCreate,
    schemas.InsuranceUpdate,
    "assets.insurance",
)
related_routes(
    "registrations",
    "asset-registrations",
    schemas.RegistrationCreate,
    schemas.RegistrationUpdate,
    "assets.registration",
)
related_routes("ownership", "asset-ownership", schemas.OwnershipCreate, None, "assets.ownership")
related_routes("media", "asset-media", schemas.MediaCreate, schemas.MediaUpdate, "assets.media")
related_routes(
    "inspections",
    "asset-inspections",
    schemas.InspectionCreate,
    schemas.InspectionUpdate,
    "assets.inspections",
)
related_routes(
    "defects", "asset-defects", schemas.DefectCreate, schemas.DefectUpdate, "assets.defects"
)
related_routes(
    "location-history", "asset-location-events", schemas.LocationEvent, None, "assets.location"
)


@router.get("/assets/{asset_id}/status-history")
async def status_history(
    asset_id: uuid.UUID,
    actor: User = Depends(require_permission("assets.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).list_records(asset_id, "status-history")


@router.post("/asset-defects/{identifier}/resolve")
async def resolve_defect(
    identifier: uuid.UUID,
    body: schemas.ResolveDefect,
    actor: User = Depends(require_permission("assets.defects.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).resolve_equipment_defect(identifier, body.notes)


@router.post("/asset-media/{identifier}/set-primary")
async def primary_media(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.media.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).primary_media(identifier)


@router.post("/asset-media/{identifier}/archive")
async def archive_media(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.media.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    return await EquipmentService(session, actor).primary_media(identifier, True)


@router.post("/assets/{asset_id}/media/upload", status_code=201)
async def upload_media(
    asset_id: uuid.UUID,
    request: Request,
    file: UploadFile = File(...),
    caption: str | None = Form(None),
    actor: User = Depends(require_permission("assets.media.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    await service.asset(asset_id, True)
    storage = request.app.state.storage
    from pathlib import Path

    if Path(file.filename or "").suffix.lower() not in {".jpg", ".jpeg", ".png"}:
        raise ValidationError("Upload a PNG or JPEG photo")
    data = await file.read(storage.max_bytes + 1)
    if not data:
        raise ValidationError("Photo is empty")
    try:
        stored = await run_in_threadpool(
            storage.save,
            f"asset-media/{actor.organization_id}/{asset_id}",
            data,
            file.filename or "",
            file.content_type,
        )
    except ValueError as error:
        raise ValidationError(str(error)) from error
    try:
        row = AssetMedia(
            organization_id=actor.organization_id,
            asset_id=asset_id,
            media_type="PHOTO",
            file_url="pending",
            storage_path=stored.relative_path,
            file_name=stored.filename,
            caption=caption,
            uploaded_by_id=actor.id,
        )
        session.add(row)
        await session.flush()
        row.file_url = f"/api/v1/asset-media/{row.id}/download"
        service.audit("asset.media_added", asset_id, row)
        await service.commit()
        return service.public(row)
    except BaseException:
        await run_in_threadpool(storage.delete, stored.relative_path)
        raise


@router.get("/asset-media/{identifier}/download")
async def download_media(
    identifier: uuid.UUID,
    request: Request,
    actor: User = Depends(require_permission("assets.media.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    row = await EquipmentService(session, actor).ref(AssetMedia, identifier)
    if not row.storage_path or not row.is_active:
        raise NotFoundError("Photo unavailable")
    path = request.app.state.storage.resolve(row.storage_path)
    if not path.is_file():
        raise NotFoundError("Photo unavailable")
    return FileResponse(path, filename=row.file_name, media_type="application/octet-stream")


@router.get("/asset-documents/{identifier}")
async def document(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.documents.read")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    return service.public(await service.ref(AssetDocument, identifier))


@router.patch("/asset-documents/{identifier}")
async def document_update(
    identifier: uuid.UUID,
    body: schemas.DocumentUpdate,
    actor: User = Depends(require_permission("assets.documents.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    row = await service.ref(AssetDocument, identifier, lock=True)
    data = body.model_dump(exclude_unset=True)
    await service.validate_fields({**service.public(row), **data}, row.asset_id)
    for key, value in data.items():
        setattr(row, key, value)
    row.verification_status = "PENDING"
    row.verified_by_id = None
    row.verified_at = None
    service.audit("asset.document_updated", row.asset_id, row)
    await service.commit()
    return service.public(row)


@router.post("/asset-documents/{identifier}/verify")
async def verify_document(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.documents.verify")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    from datetime import UTC

    service = EquipmentService(session, actor)
    row = await service.ref(AssetDocument, identifier, lock=True)
    if not row.is_active:
        raise ValidationError("Archived documents cannot be verified")
    row.verification_status = "VERIFIED"
    row.verified_by_id = actor.id
    row.verified_at = datetime.now(UTC)
    service.audit("asset.document_verified", row.asset_id, row)
    await service.commit()
    return service.public(row)


@router.post("/asset-documents/{identifier}/archive")
async def archive_document(
    identifier: uuid.UUID,
    actor: User = Depends(require_permission("assets.documents.manage")),
    session: AsyncSession = Depends(get_session),
) -> Any:
    service = EquipmentService(session, actor)
    row = await service.ref(AssetDocument, identifier, lock=True)
    row.is_active = False
    service.audit("asset.document_archived", row.asset_id, row)
    await service.commit()
    return service.public(row)


@router.get("/equipment/access")
async def equipment_access(actor: User = Depends(get_current_active_user)) -> Any:
    from app.core.dependencies import scoped_roles

    return {
        "superadmin": actor.is_superuser,
        "permissions": [p.code for r in scoped_roles(actor) for p in r.permissions],
    }
